### PR TITLE
Bind RaggedShape

### LINKED ISSUE #368: Bind RaggedShape
At some point we are going to need bindings for classs RaggedShape.  Not urgent, just putting it up here so we don't forget.

### PR DESCRIPTION
**Pull request by glynpu**
WIP
please have a review @danpovey @qindazhu  @csukuangfj 

Closes #368

### CODE REVIEW

File: k2/python/csrc/torch/ragged.cu [hunk 1]
```diff
@@ -138,6 +137,24 @@ static void PybindRaggedImpl(py::module &m) {
   PybindRaggedTpl<int32_t>(m, "RaggedInt");
 }
 
+static void PybindRaggedShape(py::module &m) {
+  py::class_<RaggedShape> pyclass(m, "RaggedShape");
+  pyclass.def("tot_size", &RaggedShape::TotSize);
+  pyclass.def("num_elements", &RaggedShape::NumElements);
```
Comment (danpovey) [inline k2/python/csrc/torch/ragged.cu:143]: Would you mind looking also at the Ragged binding and adding dim0(),  num_axes(), row_ids() and row_splits(), and to() ?

Comment (danpovey) [reply k2/python/csrc/torch/ragged.cu:143]: also max_size(), and aything else you can see that might be useful.

Comment (glynpu) [reply k2/python/csrc/torch/ragged.cu]: Finished previously mentioned functions. I am writing ragged_shape_test.py.

**Comment by danpovey** [comment]
Thanks!

File: k2/python/csrc/torch/ragged.cu [hunk 2]
```diff
@@ -141,6 +141,72 @@ static void PybindRaggedImpl(py::module &m) {
...
+        K2_CHECK(device_type == "cpu" || device_type == "cuda")
+            << "Unsupported device type: " << device_type;
+
+        ContextPtr &context = self.Context();
+        if (device_type == "cpu") {
+          if (context->GetDeviceType() == kCpu) return self;
+          return self.To(GetCpuContext());
+        }
+
+        int32_t device_index = static_cast<py::int_>(device.attr("index"));
```
Comment (csukuangfj) [inline k2/python/csrc/torch/ragged.cu:167]: Please have a look at 
https://github.com/k2-fsa/k2/blob/048da2a6f7fce1704f39a76c8dd5eef0b9b613fd/k2/python/csrc/torch/ragged.cu#L40-L43
and change `device_index` accordingly.

File: k2/python/k2/__init__.py [hunk 3]
```diff
@@ -17,6 +17,8 @@
 from .fsa_properties import to_str as properties_to_str
 from .ops import index
 from .ops import index_add
+from .ragged_shape import RaggedShape
```
Comment (csukuangfj) [inline k2/python/k2/__init__.py:20]: Please run
```bash
cd k2
flakek8 k2/python
```
and fix the style issues reported by `flake8`.

File: k2/python/tests/ragged_shape_test.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline k2/python/tests/ragged_shape_test.py:1]: Please add this file to `k2/python/tests/CMakeLists.txt`.
https://github.com/k2-fsa/k2/blob/048da2a6f7fce1704f39a76c8dd5eef0b9b613fd/k2/python/tests/CMakeLists.txt#L18-L20

File: k2/python/tests/ragged_shape_test.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#
+#  ctest --verbose -R  ragged_shape_test_py
+
+import unittest
+
+import k2
+import torch
+
+
+class TestIndex(unittest.TestCase):
```
Comment (csukuangfj) [inline k2/python/tests/ragged_shape_test.py:17]: Please change `TestIndex` to `TestRaggedShape`.

File: k2/python/tests/ragged_shape_test.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            torch.tensor([0, 0, 1, 1, 1, 2], dtype=torch.int32))
+        assert torch.allclose(
+            shape.row_ids(2),
+            torch.tensor([0, 0, 1, 2, 3, 3, 4, 5, 5, 5], dtype=torch.int32))
+        assert torch.allclose(
+            shape.row_ids(3),
+            torch.tensor([0, 0, 1, 2, 2, 3, 3, 3, 4, 5, 5, 5, 6, 7, 7, 9],
+                         dtype=torch.int32))
+
+        #TODO(liyong) test to(device)
```
Comment (csukuangfj) [inline k2/python/tests/ragged_shape_test.py:63]: Could you take a look at
https://github.com/k2-fsa/k2/blob/048da2a6f7fce1704f39a76c8dd5eef0b9b613fd/k2/python/tests/index_add_test.py#L44-L48
for how to test `to`.

All tests should run without a GPU.

File: k2/python/csrc/torch/ragged.cu (hunk 2 above)
Comment (glynpu) [reply k2/python/csrc/torch/ragged.cu]: fixed

File: k2/python/k2/__init__.py (hunk 3 above)
Comment (glynpu) [reply k2/python/k2/__init__.py:20]: fixed.

File: k2/python/tests/ragged_shape_test.py (hunk 4 above)
Comment (glynpu) [reply k2/python/tests/ragged_shape_test.py:1]: fixed.

File: k2/python/tests/ragged_shape_test.py (hunk 5 above)
Comment (glynpu) [reply k2/python/tests/ragged_shape_test.py]: fixed.

File: k2/python/tests/ragged_shape_test.py (hunk 6 above)
Comment (glynpu) [reply k2/python/tests/ragged_shape_test.py]: fixed.

**Comment by danpovey** [comment]
Cool!
@csukuangfj can you please merge if you think it's OK?

File: k2/python/tests/ragged_shape_test.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        assert shape.num_elements() >= 0
+
+        shape = k2.random_ragged_shape(False, 3, 5, 100)
+        assert shape.num_axes() >= 3
+        assert shape.num_elements() >= 100
+
+        shape = k2.random_ragged_shape(True, 3, 5, 100)
+        assert shape.num_axes() >= 3
+        assert shape.num_elements() >= 100
+        # TODO(liyong) wrap and test RaggedShape::Axes
```
Comment (csukuangfj) [inline k2/python/tests/ragged_shape_test.py:89]: Are you going to implement this todo in this pullrequest?

Comment (glynpu) [reply k2/python/tests/ragged_shape_test.py:89]: No. 
I think RaggedShapeDim is involved in test of RaggedShape::Axes. I am going to take a look at RaggedShapeDim and finish this test recently.

**Comment by csukuangfj** [comment]
+1

**Comment by csukuangfj** [comment]
I don't think RaggedShapeDim is going to be needed in Python. It's an implementation detail.



Tuesday, 24 November 2020, 21:24 +0800 from notifications@github.com  <notifications@github.com>:

**Comment by danpovey** [comment]
Yes, we won't need RaggedShapeDim wrapped.

**Comment by glynpu** [comment]
Does[ this](https://github.com/k2-fsa/k2/blob/daeaa12ad9d524d9ae896cef252a76c5ea921e1c/k2/csrc/ragged.h#L190) means  function RaggedShape::Axes should not be wrapped and needn't be tested in python level?

```
  // Axes() is intended for internal-ish use; users shouldn't really have to
  // interact with it.
  const std::vector<RaggedShapeDim> &Axes() const { return axes_; }
```

**Comment by csukuangfj** [comment]
Thanks! Merging

**Pull request merged**